src/platform/riot/uart_channel.c

```diff
@@ -0,0 +1,161 @@
+  (void)untyped_self;
+}
+
+static void UARTSyncChannel_free(NetworkChannel *untyped_self) {
+  UART_CHANNEL_DEBUG("Free");
+  (void)untyped_self;
+}
+
+static bool UARTSyncChannel_is_connected(NetworkChannel *untyped_self) {
+  UARTSyncChannel *self = (UARTSyncChannel *)untyped_self;
+  return self->state == NETWORK_CHANNEL_STATE_CONNECTED;
+}
+
+static lf_ret_t UARTSyncChannel_send_blocking(NetworkChannel *untyped_self, const FederateMessage *message) {
+  UARTSyncChannel *self = (UARTSyncChannel *)untyped_self;
+
+  if (self->state == NETWORK_CHANNEL_STATE_CONNECTED) {
+    int message_size = serialize_to_protobuf(message, self->write_buffer, UART_CHANNEL_BUFFERSIZE);
+
+    UART_CHANNEL_DEBUG("Sending Message of Size: %i", message_size);
+
+    uart_write(self->uart_dev, self->write_buffer, message_size);
+    return LF_OK;
+  } else {
+    return LF_ERR;
+  }
+}
+
+static void UARTSyncChannel_register_receive_callback(NetworkChannel *untyped_self,
+                                                      void (*receive_callback)(FederatedConnectionBundle *conn,
+                                                                               const FederateMessage *msg),
+                                                      FederatedConnectionBundle *conn) {
+  UART_CHANNEL_INFO("Register receive callback");
+  UARTSyncChannel *self = (UARTSyncChannel *)untyped_self;
+
+  self->receive_callback = receive_callback;
+  self->federated_connection = conn;
+}
+
+void _UARTSyncChannel_receive_callback(void *arg, uint8_t received_byte) {
+  UARTSyncChannel *self = (UARTSyncChannel *)arg;
+  const uint32_t minimum_message_size = 12;
+
+  self->read_buffer[self->read_index] = received_byte;
+  self->read_index++;
+
+  if (self->read_index >= minimum_message_size) {
+    if (self->super.super.category == NETWORK_CHANNEL_CATEGORY_ASYNC) {
+      cond_signal(&((UARTAsyncChannel *)self)->receive_cv);
+    }
+  }
+}
+
+void UARTSyncChannel_poll(NetworkChannel *untyped_self) {
+  UARTSyncChannel *self = (UARTSyncChannel *)untyped_self;
+
+  int bytes_left = deserialize_from_protobuf(&self->output, self->read_buffer, self->read_index);
+  UART_CHANNEL_DEBUG("Bytes Left after attempted to deserialize %d", bytes_left);
+
+  if (bytes_left >= 0) {
+    int read_index = self->read_index;
+    self->read_index = bytes_left;
+
+    memcpy(self->read_buffer, self->read_buffer + (read_index - bytes_left), bytes_left);
+
+    if (self->receive_callback != NULL) {
+      UART_CHANNEL_DEBUG("calling user callback!");
+      self->receive_callback(self->federated_connection, &self->output);
+    }
+  }
+}
+
+void *_UARTAsyncChannel_decode_loop(void *arg) {
+  UARTAsyncChannel *self = (UARTAsyncChannel *)arg;
+  mutex_lock(&self->receive_lock);
+
+  UART_CHANNEL_DEBUG("Entering decoding loop");
+  while (true) {
+    cond_wait(&self->receive_cv, &self->receive_lock);
+
+    UARTSyncChannel_poll((NetworkChannel *)&self->super);
+
+    thread_yield_higher();
```

Comment:
Why this yield here? Wont it yield when arriving at `cond_wait`?

Reply:
done